File: services/api/src/bevel_api/db/models/trace.py
"""Agent runs + append-only trace events (parallel to chat)."""

from __future__ import annotations

from datetime import datetime, timezone

from sqlalchemy import DateTime, Index, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from bevel_api.db.base import Base


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


class AgentRun(Base):
    """One agent work unit / turn — groups trace events."""

    __tablename__ = "agent_runs"
    __table_args__ = (
        Index("ix_agent_runs_tenant_room_started", "tenant_id", "room_kind", "room_id", "started_at"),
        Index("ix_agent_runs_tenant_agent_started", "tenant_id", "agent_id", "started_at"),
    )

    id: Mapped[str] = mapped_column(String(64), primary_key=True)
    tenant_id: Mapped[str] = mapped_column(String(64), nullable=False, index=True)
    # channel | agent_session | computer | cloud
    room_kind: Mapped[str] = mapped_column(String(32), nullable=False)
    room_id: Mapped[str] = mapped_column(String(256), nullable=False)
    agent_id: Mapped[str] = mapped_column(String(64), nullable=False, index=True)
    parent_run_id: Mapped[str | None] = mapped_column(String(64), nullable=True)
    message_id: Mapped[str | None] = mapped_column(String(128), nullable=True)
    # running | ok | error | cancelled
    status: Mapped[str] = mapped_column(String(16), nullable=False, default="running")
    title: Mapped[str | None] = mapped_column(String(500), nullable=True)
    started_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, default=_utcnow
    )
    ended_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    meta: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)


class AgentTraceEvent(Base):
    """One human+machine readable agent action step."""

    __tablename__ = "agent_trace_events"
    __table_args__ = (
        Index(
            "ix_trace_events_tenant_room_ts",
            "tenant_id",
            "room_kind",
            "room_id",
            "ts",
        ),
        Index("ix_trace_events_tenant_run_ts", "tenant_id", "run_id", "ts"),
        Index("ix_trace_events_tenant_agent_ts", "tenant_id", "agent_id", "ts"),
    )

    id: Mapped[str] = mapped_column(String(64), primary_key=True)
    tenant_id: Mapped[str] = mapped_column(String(64), nullable=False, index=True)
    run_id: Mapped[str] = mapped_column(String(64), nullable=False, index=True)
    room_kind: Mapped[str] = mapped_column(String(32), nullable=False)
    room_id: Mapped[str] = mapped_column(String(256), nullable=False)
    agent_id: Mapped[str] = mapped_column(String(64), nullable=False)
    message_id: Mapped[str | None] = mapped_column(String(128), nullable=True)
    # thinking | tool_call | handoff | computer | …
    kind: Mapped[str] = mapped_column(String(32), nullable=False)
    # pending | running | ok | error | cancelled
    status: Mapped[str] = mapped_column(String(16), nullable=False, default="ok")
    title: Mapped[str] = mapped_column(String(500), nullable=False)
    summary: Mapped[str | None] = mapped_column(Text, nullable=True)
    body_markdown: Mapped[str | None] = mapped_column(Text, nullable=True)
    payload: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)
    span_id: Mapped[str | None] = mapped_column(String(64), nullable=True)
    parent_span_id: Mapped[str | None] = mapped_column(String(64), nullable=True)
    ts: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, default=_utcnow, index=True
    )
    duration_ms: Mapped[int | None] = mapped_column(Integer, nullable=True)
    # none | partial | secrets_stripped
    redaction: Mapped[str] = mapped_column(String(32), nullable=False, default="none")
    schema_version: Mapped[int] = mapped_column(Integer, nullable=False, default=1)

File: packages/schema/src/trace.ts
/**
 * Agent Trace — human + machine readable agent action log (parallel to chat).
 * See docs/AGENT_TRACE.md (added with API) and plan: Agent Tracing.
 */
import { z } from 'zod'

export const TRACE_SCHEMA_VERSION = 1 as const

export const TraceRoomKindSchema = z.enum([
  'channel',
  'agent_session',
  'computer',
  'cloud',
])
export type TraceRoomKind = z.infer<typeof TraceRoomKindSchema>

export const TraceEventKindSchema = z.enum([
  'run_start',
  'run_end',
  'run_error',
  'thinking',
  'planning',
  'tool_call',
  'tool_result',
  'computer',
  'browser',
  'shell',
  'file',
  'network',
  'handoff',
  'delegate',
  'await',
  'resume',
  'model',
  'token_usage',
  'status',
  'custom',
])
export type TraceEventKind = z.infer<typeof TraceEventKindSchema>

export const TraceEventStatusSchema = z.enum([
  'pending',
  'running',
  'ok',
  'error',
  'cancelled',
])
export type TraceEventStatus = z.infer<typeof TraceEventStatusSchema>

export const TraceRedactionSchema = z.enum([
  'none',
  'partial',
  'secrets_stripped',
])
export type TraceRedaction = z.infer<typeof TraceRedactionSchema>

export const TraceRunStatusSchema = z.enum([
  'running',
  'ok',
  'error',
  'cancelled',
])
export type TraceRunStatus = z.infer<typeof TraceRunStatusSchema>

/** One agent action step — dual human/machine payload. */
export const AgentTraceEventSchema = z.object({
  id: z.string().min(1).max(64),
  schemaVersion: z.literal(TRACE_SCHEMA_VERSION).default(TRACE_SCHEMA_VERSION),
  tenantId: z.string().min(1),
  runId: z.string().min(1).max(64),
  parentRunId: z.string().max(64).optional(),
  spanId: z.string().max(64).optional(),
  parentSpanId: z.string().max(64).optional(),

  roomKind: TraceRoomKindSchema,
  roomId: z.string().min(1).max(256),
  messageId: z.string().max(128).optional(),
  agentId: z.string().min(1).max(64),
  actor: z.string().max(320).optional(),

  ts: z.string().datetime({ offset: true }).or(z.string().datetime()),
  durationMs: z.number().int().nonnegative().optional(),

  kind: TraceEventKindSchema,
  title: z.string().min(1).max(500),
  summary: z.string().max(4000).optional(),
  bodyMarkdown: z.string().max(32_000).optional(),

  payload: z.record(z.string(), z.unknown()).default({}),
  status: TraceEventStatusSchema.default('ok'),
  redaction: TraceRedactionSchema.default('none'),
})
export type AgentTraceEvent = z.infer<typeof AgentTraceEventSchema>

/** Ingest body: client may omit id/ts/schemaVersion (server fills). */
export const AgentTraceEventIngestSchema = AgentTraceEventSchema.partial({
  id: true,
  schemaVersion: true,
  ts: true,
  status: true,
  redaction: true,
  payload: true,
}).required({
  tenantId: true,
  runId: true,
  roomKind: true,
  roomId: true,
  agentId: true,
  kind: true,
  title: true,
})
export type AgentTraceEventIngest = z.infer<typeof AgentTraceEventIngestSchema>

export const AgentTraceBatchIngestSchema = z.object({
  events: z.array(AgentTraceEventIngestSchema).min(1).max(100),
})
export type AgentTraceBatchIngest = z.infer<typeof AgentTraceBatchIngestSchema>

export const AgentRunSchema = z.object({
  id: z.string().min(1).max(64),
  tenantId: z.string().min(1),
  roomKind: TraceRoomKindSchema,
  roomId: z.string().min(1).max(256),
  agentId: z.string().min(1).max(64),
  parentRunId: z.string().max(64).optional(),
  messageId: z.string().max(128).optional(),
  status: TraceRunStatusSchema,
  title: z.string().max(500).optional(),
  startedAt: z.string(),
  endedAt: z.string().optional(),
  meta: z.record(z.string(), z.unknown()).default({}),
})
export type AgentRun = z.infer<typeof AgentRunSchema>

export const AgentRunOpenSchema = z.object({
  id: z.string().min(1).max(64).optional(),
  tenantId: z.string().min(1),
  roomKind: TraceRoomKindSchema,
  roomId: z.string().min(1).max(256),
  agentId: z.string().min(1).max(64),
  parentRunId: z.string().max(64).optional(),
  messageId: z.string().max(128).optional(),
  title: z.string().max(500).optional(),
  meta: z.record(z.string(), z.unknown()).optional(),
})
export type AgentRunOpen = z.infer<typeof AgentRunOpenSchema>

export const AgentRunCloseSchema = z.object({
  status: TraceRunStatusSchema.exclude(['running']),
  title: z.string().max(500).optional(),
  meta: z.record(z.string(), z.unknown()).optional(),
})
export type AgentRunClose = z.infer<typeof AgentRunCloseSchema>
